### DISCUSSION TITLE
Why is behavior of machine.withContext and machine.withConfig different?

### DISCUSSION THREAD

**Discussion by danielkcz** [statelyai/xstate#1844] [Q&A] [answered]
What I mean is that `withContext` always overrides the initial context while `withConfig({ context: {...}})` will merge it on top of the initial one. In more elaborate systems such an oversight causes unexpected behavior that's rather hard to track.

Furthermore, I haven't found in docs any mention of either behavior. The [Quick reference](https://xstate.js.org/docs/guides/context.html#quick-reference) in docs should probably mention `withConfig` alternative as well explaining how is it different from `withContext`.

**Comment by Andarist** [comment, marked as answer]
This looks like inconsistency to me - fixing this right now would be a breaking change that could affect many users. As such I believe that it's best to report an issue about it so we can evaluate how it should work and change that in v5.

**Reply by danielkcz** [reply]
Hmm, looks like I am wrong actually. The `withConfig` does not support passing `context` at all. I found a small mention of that here: https://github.com/davidkpiano/xstate/blob/87d0acd9e9530079829ad30228558b18b77ea4a2/packages/xstate-react/CHANGELOG.md#added, verified and it's true. I did the wrong test previously and thought it's merging.

Anyway, I found [this page in docs](https://xstate.js.org/docs/guides/machines.html#extending-machines) which explains that `withContext` is not merging and how to deal with partial context. I suppose that will suffice for now.

Oddly enough, when I search docs for `withContext` it won't find that page, but when I do search for `withConfig` it finds that.

**Comment by rlaffers** [comment]
Actually, there is an inconsistency; I'm not sure if it is intentional though:
1. machine.withContext: overrides the entire context.
2. useMachine(machine, {
    context: { value: 42 }, // merges with the current machine context
  }
  
I personally find the merge behavior more convenient. Yes, these two cases are technically different (extending a machine vs extending a service), but they both boil down to providing context to the machine dynamically. Just out of curiosity, is this intentional? If so, what is the reasoning behind it? Thanks.

**Reply by danielkcz** [reply]
Are you sure about that merge? I thought that too, but from what I tried, it was actually ignored. Perhaps set up a small repro?

**Reply by rlaffers** [reply]
Sure, here you go https://codesandbox.io/s/condescending-kowalevski-z48m2

**Reply by danielkcz** [reply]
Oh right, I misread your initial information and thought you are talking about `withConfig`. In that case, it's best to open the issue as Andarist said above to track it. It will have to be a breaking change in V5 to ensure consistency.

**Reply by rlaffers** [reply]
#1882
